### PR TITLE
feat(weave): Optimize publish speed

### PR DESCRIPTION
**Pull request by vwrj**
JIRA: https://wandb.atlassian.net/browse/WB-16217
Picking low-hanging fruits from doc: https://www.notion.so/wandbai/Weave-Publish-speed-investigation-ae33f479fe4e41469fc7a3e593583778

Chops off ~1.1 seconds from publish time. 
- Remove second lite_run.run access (after finishing the LiteRun -- this needlessly re-executes queries such as UpsertModel, UpsertBucket, etc).
- Extends responsibility of `InMemoryLazyLiteRun` to include logging an artifact and exposing that as a method to callers. 
- Adds a boolean `artifact_collection_exists` that's plumbed through the commit/publish path to ensure that we don't create an artifact type if the artifact collection already exists
- Adds a `setup_log_deps(...)` to `InMemoryLazyLiteRun` that lazily executes a FetchHistoryStep query (and relevant logic for logging) that is not needed in the artifact or run creation paths.
- Adds an `upsert_project(...)` method to `InMemoryLazyLiteRun` that ensures that project creation only happens once when called in the lifetime of a class instance.

### CODE REVIEW

File: weave/wandb_interface/wandb_artifact_pusher.py [hunk 1]
```diff
@@ -92,8 +97,8 @@ def write_artifact_to_wandb(
 
     # Return the URI of the artifact
     return WeaveWBArtifactURIComponents(
-        entity_name=lite_run.run.entity,
```
Comment (vwrj) [inline weave/wandb_interface/wandb_artifact_pusher.py:95]: This actually triggers a duplicate cascade of UpsertModel, UpsertBucket, etc. because the `lite_run.finish()` call above sets _run in the InMemoryLazyLiteRun class to None.

File: weave/wandb_interface/wandb_lite_run.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ def __init__(
         group: typing.Optional[str] = None,
         _hide_in_wb: bool = False,
         _use_async_file_stream: bool = False,
+        _optimize_for_artifact_commit: bool = False,
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_lite_run.py:54]: `optimize_for_artifact_commit` as a variable name implies that the lite run knows what is needed to optimize for artifact committing. Feels like a bleed of responsibilities. I think this could be modified with 2 changes:
1. it might be called: `_create_project_if_needed` or something like that and default to `true`. Which will do exactly what the variable says
2. The step initialization:
```
            self._step = 0
            if not self._optimize_for_artifact_commit:
                self._step = self._run.lastHistoryStep + 1
```
can be refactored to lazily initialize when needed. This will remove the dependency on this constructor param.

File: weave/wandb_interface/wandb_artifact_pusher.py (hunk 1 above)
Comment (tssweeney) [reply weave/wandb_interface/wandb_artifact_pusher.py:95]: nice find

File: weave/wandb_interface/wandb_artifact_pusher.py [hunk 3]
```diff
@@ -46,17 +47,25 @@ def write_artifact_to_wandb(
...
+            entity_name,
+            project_name,
+            group="weave_artifact_pushers",
+            _hide_in_wb=True,
+            _optimize_for_artifact_commit=True,
         )
     else:
         lite_run = _lite_run
 
+    entity_name = lite_run.run.entity or entity_name
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_artifact_pusher.py:59]: shouldn't entity_name and project_name always be valid here?

File: weave/wandb_interface/wandb_artifact_pusher.py [hunk 4]
```diff
@@ -28,6 +28,7 @@ def write_artifact_to_wandb(
     additional_aliases: list = [],
     *,
     _lite_run: typing.Optional[InMemoryLazyLiteRun] = None,
+    merge: typing.Optional[bool] = False,
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_artifact_pusher.py:31]: 2 comments: 
1. `merge` is a confusing param name because it implies that `write_artifact_to_wandb` knows what it means to "merge" something. I think it would be cleaner to be explicit: `assume_artifact_type_exists`.
2. stylistically, it would be better to put non-underscore params before underscore ones.

File: weave/wandb_interface/wandb_lite_run.py (hunk 2 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_lite_run.py:54]: I don't think the lazy init on history step works here. When the Filepusher is initialized, it also initializes Filestream which uses the history step with the following code:

```python
self._stream.set_file_policy(
    "wandb-history.jsonl",
    file_stream.JsonlFilePolicy(start_chunk_id=**self._step**),
)
```

We need to disambiguate which context LiteRun is used under. So I think the boolean flag is the way to go. 

I'll do two boolean flags which both default to true:
- `create_project`
- `get_last_history_step`

File: weave/wandb_interface/wandb_artifact_pusher.py (hunk 3 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_artifact_pusher.py:59]: Calling `.run` here kicks off the UpsertProject mutation. This code path is actually used when publishing brand new boards / creating brand new artifacts under projects don't yet exist. If we don't kick off the UpsertProject here, we have a bunch of tests that fail at the next line (CreateArtifactType) since the project doesn't exist. Which is why I assume the previous code called `lite_run.run.entity` and `lite_run.run.project`. 

I'm not a fan of this happening implicitly. So I'm actually going to place a `upsert_project` method on the LiteRun class that we can call explicitly so I can change this back to use the passed-in entity and project names.

File: weave/wandb_interface/wandb_artifact_pusher.py (hunk 4 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_artifact_pusher.py:31]: done

File: weave/artifact_wandb.py [hunk 5]
```diff
@@ -647,6 +647,7 @@ def save(
         branch: typing.Optional[str] = None,
         *,
         _lite_run: typing.Optional["InMemoryLazyLiteRun"] = None,
+        _create_new_artifact: typing.Optional[bool] = True,
```
Comment (tssweeney) [inline weave/artifact_wandb.py:650]: sorry to double nit, but shouldn't this be `create_new_artifact_type`?

Comment (vwrj) [reply weave/artifact_wandb.py:650]: I think it makes more sense to have a boolean that tells us what path we're in: whether we're creating a brand new artifact (publish_new, publish_as) or creating an artifact version inside of it (commit). The default case is a brand new artifact in which case we will upsert the project, create artifact types, do the run history keys query, etc.

File: weave/wandb_interface/wandb_lite_run.py [hunk 6]
```diff
@@ -75,6 +77,10 @@ def __init__(
             _use_async_file_stream
             and os.getenv("WEAVE_DISABLE_ASYNC_FILE_STREAM") == None
         )
+        self._fetch_last_history_step = _fetch_last_history_step
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_lite_run.py:80]: this should always be equal to `not _use_async_file_stream` since that is the only time step is used.

Comment (vwrj) [reply weave/wandb_interface/wandb_lite_run.py:80]: `_use_async_file_stream` is set to false here by default. Which means in our commit case, it would always fetch the last history step and make the `RunHistoryKeys` query.

Comment (tssweeney) [reply weave/wandb_interface/wandb_lite_run.py:80]: this still is bleeding internal implementation details. In general, you want the parameters of something to change the effect it has, not how it works. Here, fetching history step is an internal piece of logic. I don't think the caller should be setting this flag. The implementation itself should be smart enough to only fetch the step once and only if needed to perform the operation where it is needed.

File: weave/wandb_interface/wandb_artifact_pusher.py [hunk 7]
```diff
@@ -46,17 +49,25 @@ def write_artifact_to_wandb(
 
     if _lite_run is None:
         lite_run = InMemoryLazyLiteRun(
-            entity_name, project_name, group="weave_artifact_pushers", _hide_in_wb=True
+            entity_name,
+            project_name,
+            group="weave_artifact_pushers",
+            _hide_in_wb=True,
+            _always_create_project=_create_new_artifact,
+            _fetch_last_history_step=_create_new_artifact,
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_artifact_pusher.py:57]: Why do you want to fetch the last history step when you are creating a new artifact? I would imagine this should always be False here

File: weave/wandb_interface/wandb_artifact_pusher.py [hunk 8]
```diff
@@ -46,17 +49,25 @@ def write_artifact_to_wandb(
...
     else:
         lite_run = _lite_run
 
-    # Ensure the artifact type exists
-    lite_run.i_api.create_artifact_type(
-        artifact_type_name=artifact_type_name,
-        entity_name=lite_run.run.entity,
-        project_name=lite_run.run.project,
-    )
+    lite_run.upsert_project()
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_artifact_pusher.py:62]: Why is this needed? I think this always runs anyway in your implementation

File: weave/wandb_interface/wandb_lite_run.py [hunk 9]
```diff
@@ -51,6 +51,8 @@ def __init__(
         group: typing.Optional[str] = None,
         _hide_in_wb: bool = False,
         _use_async_file_stream: bool = False,
+        _always_create_project: bool = True,
```
Comment (tssweeney) [inline weave/wandb_interface/wandb_lite_run.py:54]: the term `always` is not clear to me - what does it mean to set this to false?

File: weave/wandb_interface/wandb_lite_run.py (hunk 6 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_lite_run.py:80]: This class is used in two different contexts: for StreamTables and to upload artifacts. We are using it for the latter. How will the class know which context it's instantiated in? In our context, we _never_ need to fetch the history keys. And we _never_ need to upsert a project. These are queries that should never happen in our context, not even once. Hence why I created underscored boolean flags as arguments. What's the proposed solution?

File: weave/wandb_interface/wandb_artifact_pusher.py (hunk 7 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_artifact_pusher.py:57]: Creating a brand new artifact (collection) is the default path and for that, we need to make sure that a project is upserted. But fetching the last history step is not necessary here.  Agreed, I'll change it to False!

File: weave/wandb_interface/wandb_artifact_pusher.py (hunk 8 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_artifact_pusher.py:62]: I'm not a fan how `lite_run.run.entity` caused an upsert project query to happen. The fact that it was running in the background implicitly made it hard to detect. So I decided to refactor to make it explicit.

File: weave/wandb_interface/wandb_lite_run.py (hunk 9 above)
Comment (vwrj) [reply weave/wandb_interface/wandb_lite_run.py:54]: fair -- I can remove `always`.

File: weave/wandb_interface/wandb_artifact_pusher.py (hunk 8 above)
Comment (tssweeney) [reply weave/wandb_interface/wandb_artifact_pusher.py:61]: Ok, i see. It strikes me as odd to have `should_create_project=_create_new_artifact` then call `lite_run.upsert_project()` immediately after. 

As a user of this API, how do I know when i need to call `upsert_project`? Why is that not covered with `should_create_project=_create_new_artifact`? I think you might be able to refactor this a bit more to collocate the related calls.

Comment (vwrj) [reply weave/wandb_interface/wandb_artifact_pusher.py:61]: I can understand why you'd expect `should_create_project` to take care of upserting the project. However, this boolean is only being used to create the project when creating the run. In this case, we need `upsert_project` to be called explicitly because the `lite_run.i_api.create_artifact_type` call below requires that the project always exists. If we don't explicitly call `upsert_project`, it actually fails some tests/code paths (i.e when publishing a brand new artifact to wandb to a project that doesn't exist yet). 

I can rename the boolean to `create_project_for_run` if you'd like, but that would leak internal implementation details to the caller.

**Review by tssweeney — APPROVED** [review]
Amazing! Looks great (: Thank you for the refactor
